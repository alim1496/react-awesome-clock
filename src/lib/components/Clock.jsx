import React, { useEffect, useState } from "react"
import PropTypes from "prop-types"
import "./clock.css"
import HourMark from "./HourMark"
import Hand from "./Hand"

const Clock = (props) => {
	const { hourValue, minuteValue, size = 200, borderColor = "#000", borderWidth = 5, hourFormat, live } = props
	const hours = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]
	const romans = ["", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX", "X", "XI", "XII"]

	const [hour, setHour] = useState(0)
	const [minute, setMinute] = useState(0)
	const [second, setSecond] = useState(0)

	useEffect(() => {
		if (live) {
			const interval = setInterval(() => {
				const date = new Date()
				const hh = date.getHours()
				const mm = date.getMinutes()
				const ss = date.getSeconds()
				setHour(hh)
				setMinute(mm)
				setSecond(ss)
			}, 1000)
			return () => clearInterval(interval)
		} else {
			setMinute(minuteValue)
			setHour(hourValue)
		}
	}, [live, minuteValue, hourValue])

	return (
		<div style={{ width: size, height: size, borderColor, borderWidth }} className="rac-clock-body">
			{hours.map((hour, index) => (
				<HourMark
					key={index} 
					value={hourFormat === "number" ? hour : romans[hour]} 
					angle={hour * 30} 
					format={hourFormat} 
				/>
			))}
			<div className="rac-hand-wrapper" style={{ width: size, height: size/2 }}>
				{live && <Hand height={4*size/10} handID="rac-second-hand" angle={second*6} />}
				<Hand height={3*size/10} handID="rac-minute-hand" angle={minute*6+second/10} />
				<Hand height={size/5} handID="rac-hour-hand" angle={hour*30+minute/2} />
			</div>
		</div>
	)
}

Clock.propTypes = {
	hourValue: PropTypes.number,
	minuteValue: PropTypes.number,
	size: PropTypes.number,
	borderColor: PropTypes.string,
	borderWidth: PropTypes.number,
	hourFormat: PropTypes.string,
	live: PropTypes.bool
}

export default Clock