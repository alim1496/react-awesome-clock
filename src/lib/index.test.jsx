import React from "react"
import renderer from "react-test-renderer"
import { describe, expect, test } from "vitest"
import Clock from "./index"

describe("Clock", () => {
	test("Clock component renders correctly", () => {
		const clock = renderer.create(<Clock />)
		const tree = clock.toJSON()
		expect(tree).toBeTruthy()
	})
})