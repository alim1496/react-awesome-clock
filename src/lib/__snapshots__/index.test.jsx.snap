// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Clock > Clock component renders correctly 1`] = `
<div>
  This is a clock
</div>
`;
